### ISSUE TITLE
Python 3.7 

### ISSUE THREAD

**Issue by Saremox** [spack/spack#8968]
Summary

Freshly cloned develop branch throws errors on Python 3.7

### Expected Result

Packages being installed

### Actual Result

```
:~/Repos % git clone https://github.com/spack/spack.git
Cloning into 'spack'...
remote: Counting objects: 122158, done.
remote: Compressing objects: 100% (110/110), done.
remote: Total 122158 (delta 68), reused 80 (delta 24), pack-reused 122007
Receiving objects: 100% (122158/122158), 41.35 MiB | 2.62 MiB/s, done.
Resolving deltas: 100% (59189/59189), done.
:~/Repos % cd spack/bin 
:~/Repos/spack/bin (develop=) % ./spack install libelf
==> Error: name 'AutotoolsPackage' is not defined
```

### Information on your system

```
Archlinux x64
:~/Repos/spack/bin (develop=) % python --version
Python 3.7.0
:~/Repos/spack/bin (develop=) % uname -a
Linux 4.17.12-pascal #1 SMP PREEMPT Sun Aug 5 15:51:19 UTC 2018 x86_64 GNU/Linux
```
### How To reproduce in a docker container

```
:~ # docker run -i base/archlinux bash 
pacman -Sy git python   
:: Synchronizing package databases...
[...]
Packages (7) libnsl-1.2.0-1  pcre2-10.31-1  perl-error-0.17026-2  perl-mailtools-2.20-2  perl-timedate-2.30-5  git-2.18.0-1  python-3.7.0-3

Total Download Size:    41.15 MiB
Total Installed Size:  182.78 MiB

:: Proceed with installation? [Y/n] 
[...]
git clone https://github.com/spack/spack.git
Cloning into 'spack'...
cd spack/bin
./spack install libelf
==> Error: name 'AutotoolsPackage' is not defined
```

**Comment by adamjstewart** [comment]
**TL;DR:** Run `spack clean --python-cache` to permanently fix this problem.

This bug was introduced by #8090, which moved around a lot of Spack's internals to make it faster. For Python 2, no one noticed any problems, but Python 3 is more aggressive about caching. What happens is that Spack tries to load the caches from the old packages, but since the imports all moved around, they no longer work. Deleting all of the `__pycache__` and `*.pyc` files solves the problem.

Duplicate of #8292 
Duplicate of #8248 
Duplicate of #8413

Also see #8419 and #8090

**Comment by Saremox** [comment]
The behavior is reproducible even on a clean OS install, see my attached sequence in a docker container above.

Nonetheless i gave your suggestion a try.
```
:~/Repos/spack (develop=) % bin/spack clean --python-cache
==> Removing python cache files
:~/Repos/spack (develop=) % find -name "*pyc"           
:~/Repos/spack (develop=) % bin/spack install libelf      
==> Error: name 'AutotoolsPackage' is not defined
```
Same outcome as above. This wasn't the case with Python 3.6.x and only occurred after upgrading to 3.7

**Comment by adamjstewart** [comment]
@tgamblin?

**Comment by Saremox** [comment]
I have done some testing with different Distributions and python version. 

| Distribution / Python version | 3.5.3 | 3.6.6 | 3.7 |
|---|---|---|---|
| Archlinux | | :heavy_check_mark:  | :x: |
| Debian buster | | :x: | :x:  |
| Debian Stretch | :heavy_check_mark: | | |
| Ubuntu bionic | :soon: | :soon: | :soon: |

### Debian Buster Python 3.6.6
```
:~ # docker run -i debian:buster bash 
apt-get update
apt-get install make curl git python3 gcc
git clone https://github.com/spack/spack.git
cd spack/bin
python3 ./spack install libelf
==> Error: No module named 'distutils.dir_util'

```
Trying to clean cache:
```
python3 ./spack clean --python-cache
==> Removing python cache files
python3 ./spack install libelf
==> Error: No module named 'distutils.dir_util'
python3 --version
Python 3.6.6
```
### Debian Buster - Python 3.7

```
apt-get install make curl git python3.7 gcc
python3.7 ./spack clean --python-cache
==> Removing python cache files
python3.7 ./spack install libelf
==> Error: No module named 'distutils.dir_util'
```

### Debian Stretch - Python 3.5.3
```
apt-get update
apt-get install make curl git python3 gcc
git clone https://github.com/spack/spack.git
cd spack/bin
==> Installing libelf
==> Using cached archive: /spack/var/spack/cache/libelf/libelf-0.8.13.tar.gz
==> Staging archive: /spack/var/spack/stage/libelf-0.8.13-xw46i6lvhazk7diredfvwakjrsngc4aj/libelf-0.8.13.tar.gz
==> Created stage in /spack/var/spack/stage/libelf-0.8.13-xw46i6lvhazk7diredfvwakjrsngc4aj
==> No patches needed for libelf
==> Building libelf [AutotoolsPackage]
==> Executing phase: 'autoreconf'
==> Executing phase: 'configure'
==> Executing phase: 'build'
==> Executing phase: 'install'
==> Successfully installed libelf
  Fetch: 0.00s.  Build: 3.34s.  Total: 3.34s.
[+] /spack/opt/spack/linux-debian9-x86_64/gcc-6.3.0/libelf-0.8.13-xw46i6lvhazk7diredfvwakjrsngc4aj
```

**Comment by adamjstewart** [comment]
Hmm, that's a completely different error message. I thought [distutils.dir_util](https://docs.python.org/3.7/distutils/apiref.html#module-distutils.dir_util) was a core Python module. Is it possible to install Python without distutils?

**Comment by Saremox** [comment]
I looked into that and for some reason with Debian buster they decided to put distutils into it's own package. And since bionic is allrdy based on debian buster the error appears there too.

I have redone the above tests now with installed distutils

| Distribution / Python version | 3.5.x | 3.6.x | 3.7 |
|---|---|---|---|
| Archlinux | | :heavy_check_mark:  | :x: |
| Debian buster | | :heavy_check_mark: | :heavy_check_mark:  |
| Debian Stretch | :heavy_check_mark: | | |
| Ubuntu bionic |  | :heavy_check_mark: | :heavy_check_mark:  |

I'll report that on the archlinux bug tracker it now looks like a problem with the packaged python version and not Python 3.7. https://bugs.archlinux.org/task/59650

**Comment by scheibelp** [comment]
Thanks for investigating this so thoroughly! Would you be able to re-run the command that fails on Archlinux with `-sd`: `bin/spack -sd install libelf` (the options go before the `install` command).

**Comment by Saremox** [comment]
```
:~/Repos/spack (develop=) % bin/spack -sd install libelf
/home/michael/Repos/spack/lib/spack/spack/config.py:630 ==> Reading config file /home/michael/Repos/spack/etc/spack/defaults/modules.yaml
/home/michael/Repos/spack/lib/spack/spack/config.py:630 ==> Reading config file /home/michael/Repos/spack/etc/spack/defaults/config.yaml
/home/michael/Repos/spack/lib/spack/spack/config.py:630 ==> Reading config file /home/michael/Repos/spack/etc/spack/defaults/repos.yaml
Traceback (most recent call last):
  File "bin/spack", line 61, in <module>
    sys.exit(spack.main.main())
  File "/home/michael/Repos/spack/lib/spack/spack/main.py", line 653, in main
    return _invoke_command(command, parser, args, unknown)
  File "/home/michael/Repos/spack/lib/spack/spack/main.py", line 432, in _invoke_command
    return_val = command(parser, args)
  File "/home/michael/Repos/spack/lib/spack/spack/cmd/install.py", line 214, in install
    args.package, concretize=True, tests=tests)
  File "/home/michael/Repos/spack/lib/spack/spack/cmd/__init__.py", line 163, in parse_specs
    spec.concretize(tests=tests)  # implies normalize
  File "/home/michael/Repos/spack/lib/spack/spack/spec.py", line 1845, in concretize
    user_spec_deps=user_spec_deps),
  File "/home/michael/Repos/spack/lib/spack/spack/spec.py", line 2239, in normalize
    self.validate_or_raise()
  File "/home/michael/Repos/spack/lib/spack/spack/spec.py", line 2291, in validate_or_raise
    spack.repo.get(spec.fullname)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 1125, in get
    return path.get(spec)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 101, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 577, in get
    return self.repo_for_pkg(spec).get(spec)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 101, in converter
    return function(self, spec_like, *args, **kwargs)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 831, in get
    package_class = self.get_pkg_class(spec.name)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 1025, in get_pkg_class
    module = self._get_pkg_module(pkg_name)
  File "/home/michael/Repos/spack/lib/spack/spack/repo.py", line 999, in _get_pkg_module
    module = imp.load_source(fullname, file_path, f)
  File "/usr/lib/python3.7/imp.py", line 171, in load_source
    module = _load(spec)
  File "<frozen importlib._bootstrap>", line 696, in _load
  File "<frozen importlib._bootstrap>", line 677, in _load_unlocked
  File "<frozen importlib._bootstrap_external>", line 728, in exec_module
  File "<frozen importlib._bootstrap>", line 219, in _call_with_frames_removed
  File "/home/michael/Repos/spack/var/spack/repos/builtin/packages/libelf/package.py", line 28, in <module>
    class Libelf(AutotoolsPackage):
NameError: name 'AutotoolsPackage' is not defined
```

**Comment by tgamblin** [comment]
I just merged #8289, and I don't think the Spack core uses `distutils` anymore.  Does that at least fix the `distutils` problems here?

**Comment by adamjstewart** [comment]
@tgamblin The only remaining place in Spack that uses `distutils` is the Python package. It is used to locate the Python home, libs, and headers, and to set up `PYTHONHOME`. So anything that depends on Python will not work with an external Python that does not have `distutils` installed.

**Comment by tgamblin** [comment]
@adamjstewart yeah I saw that, but as you say, it's executed by the `python` Spack's *installing* into.  If there's anything left in the core that invokes `distutils`, we should probably get rid of it.

In general I'm not sure how much distribution hackery we can really tolerate.  I hope distros don't rip out other parts of the Python core.

**Comment by adamjstewart** [comment]
@tgamblin Or by the `python` Spack is installing extensions for. If I add my system-installed `python` as an external package, it will try to import `distutils` using that `python` installation. The obvious workaround being "let Spack install Python for you".

**Comment by tgamblin** [comment]
@Saremox: can you verify that things work without `distutils` now?  Still looking at the other issue.

**Comment by healther** [comment]
> The obvious workaround being "let Spack install Python for you".

Maybe this should just go into the FAQ or known issues section

**Comment by tgamblin** [comment]
Ok it looks like this is an issue with `imp.load_source()`.  We call it like this:

```python
with prepend_open(file_path, text=_package_prepend) as f:
    module = imp.load_source(fullname, file_path, f)
```

The outer context handler prepends `from spack.pkgkit import *` to all packages, and it returns a file handle (the `file` argument to `load_source` annoyingly has to be a low-level file handle).  `load_source` then loads the modified package from `f`, but we pass the `file_path` of the original python file in the repository so that the resulting module things that is where it's located.  That's where the module's `__file__` attribute comes from, which is how Spack finds things like patches that live in the same directory as the package.

On arch, when we import `libelf/package.py`, it's importing the regular version instead of the version with the import prepended.  So it seems like the 3.7 implementation on `arch` ignores the `file` argument to `load_source` and just opens the path it's given.  I'm not sure why I don't see the same bug on other platforms, though...  I think we may need to add some support to use `importlib` instead of `imp` to get down low enough to fix this.

**Comment by adamjstewart** [comment]
> I think we may need to add some support to use `importlib` instead of `imp` to get down low enough to fix this.

That's probably a good idea. According to https://docs.python.org/3/library/imp.html:

> _Deprecated since version 3.4:_ The `imp` package is pending deprecation in favor of `importlib`.

**Comment by tgamblin** [comment]
> That's probably a good idea. According to https://docs.python.org/3/library/imp.html:
> 
> > _Deprecated since version 3.4:_ The `imp` package is pending deprecation in favor of `importlib`.

Yes.  The issue is really that `importlib` doesn't exist in 2.6 😭.

**Comment by adamjstewart** [comment]
Still dreaming of the day when Spack moves to PyPI and I can `pip install spack` and its growing list of vendored dependencies. But we would probably have so many that everything would break as soon as a new version was released. Guess we'll have to vendor [importlib](https://pypi.org/project/importlib/) 😄

**Comment by tgamblin** [comment]
`imp` seems to work fine for 2.7 -- so I think in this instance we can just use one or the other conditionally.

**Comment by tgamblin** [comment]
@Saremox @adamjstewart: can you try out #8995?  I think it fixes the issue with 3.7 on arch (by using `importlib` if it's available).

Imports should also be faster in python 3 (no more tempfiles).
